pandas/_libs/lib.pyx

```diff
@@ -581,6 +581,13 @@ def array_equivalent_object(left: object[:], right: object[:]) -> bool:
                     return False
             elif (x is C_NA) ^ (y is C_NA):
                 return False
+            # Only compare scalars to scalars and arrays to arrays
+            elif cnp.PyArray_IsAnyScalar(x) != cnp.PyArray_IsAnyScalar(y):
```

Comment:
can you put these comments indented inside the conditions

Reply:
i think id rather use our `is_scalar` than `cnp.PyArray_IsAnyScalar`

Reply:
If I change it to `is_scalar` one unit test fails and the perf decreases even more. It is almost twice as slow as using `cnp.PyArray_IsAnyScalar`

       before           after         ratio
     [822dc6f9]       [eeabb266]
                      <gh35267>
+        25.6±1ms         99.8±2ms     3.90  frame_methods.Equals.time_frame_nonunique_unequal
+        28.2±3ms         97.3±1ms     3.44  frame_methods.Equals.time_frame_nonunique_equal
+        31.2±2ms        106±0.9ms     3.40  frame_methods.Equals.time_frame_object_unequal
+        43.3±3ms        112±0.7ms     2.59  frame_methods.Equals.time_frame_object_equal

Reply:
Put comments indented inside the conditons.

Reply:
im pre-caffeine so bear with me: what does PyArray_IsAnyScalar include?  or more specifically, what does it _not_ include that is_scalar does?

> is_scalar one unit test fails

what breaks?

Reply:
`PyArray_IsAnyScalar` does not include None,  datetime.datetime, datetime.timedelta, Period, decimal.Decimal, Fraction, numbers.Number, Interval DateOffset, while `is_scalar` does.

If I see this I think it would be good to use `is_scalar` for consistency. But the drawback is the worse perf, because under the hood `is_scalar` makes use of `PyArray_IsAnyScalar` and some other functions.

Nvm the failed unit test, I see that it is a broken test on master that appeared after I merged master (`pandas/tests/io/test_parquet.y::test_s3_roundtrip_for_dir)`

Reply:
@jbrockmendel how do we want to continue with this? Shall I replace  `PyArray_IsAnyScalar` with `is_scalar`, or leave it as it is?

Reply:
IIRC `not is_listlike(obj)` is more general and more performant than `is_scalar(obj)`.  would that work in this context?

Reply:
If I replace `PyArray_IsAnyScalar` with `is_listlike` and update the code like below, the perf becomes (much) worse. Will try the try/except option jreback proposed.

``` python
elif is_list_like(x) != is_list_like(y):
    # Only compare scalars to scalars and non-scalars to non-scalars
    return False
elif (is_list_like(x) and is_list_like(y)
       and not (isinstance(x, type(y)) or isinstance(y, type(x)))):
    # Check if non-scalars have the same type
   return False
```

       before           after         ratio
     [822dc6f9]       [98cea4b3]
     <master~31>       <gh35267_clean>
+      25.1±0.6ms          695±2ms    27.65  frame_methods.Equals.time_frame_nonunique_equal
+      25.4±0.6ms          702±3ms    27.65  frame_methods.Equals.time_frame_nonunique_unequal
+      34.0±0.3ms          711±9ms    20.92  frame_methods.Equals.time_frame_object_unequal
+      44.5±0.5ms          718±3ms    16.14  frame_methods.Equals.time_frame_object_equal